Make the RAND code available from within the FIPS module

mattcaswell opened the pull request:
This enables calls to RAND_*() functions from within the FIPS module. This is required in order to bring in support for BIGNUM, and the various asymmetric algorithms.

I've added a dummy call into the FIPS module just to demonstrate that it works. This will be removed in due course when we no longer need it.

slontis commented on crypto/async/async.c:
```diff
@@ -374,7 +376,8 @@ int ASYNC_init_thread(size_t max_size, size_t init_size)
 }
 
-void async_delete_thread_state(void)
+/* OPENSSL_CTX ignored for now */
```
TODO 3.0 ?

slontis commented on crypto/context.c:
```diff
@@ -155,11 +169,7 @@ static int openssl_ctx_init_index(OPENSSL_CTX *ctx, int static_index,
-            return 0;
-        ctx = default_context;
-    }
+    ctx = openssl_ctx_get_concrete(ctx);
```
Do we want this kind of pattern? Or would it be better to just call the function always and 
#ifdef FIPS_MODE
# define openssl_ctx_get_concrete(ctx) ctx
#endif

slontis commented on crypto/err/err.c:
```diff
@@ -685,7 +685,8 @@ const char *ERR_reason_error_string(unsigned long e)
 }
 
-void err_delete_thread_state(void)
+/* OPENSSL_CTX ignored for now */
```
TODO 3.0

slontis commented on crypto/include/internal/cryptlib_int.h:
```diff
@@ -7,17 +7,16 @@
+                           OSSL_thread_stop_handler_fn handfn);
+int init_thread(void);
+void cleanup_thread(void);
+void ossl_ctx_thread_stop(void *arg);
```
These start/stop names look different

mattcaswell replied:
Not sure I understand your point. There isn't a "start" equivalent of `ossl_ctx_thread_stop`

levitte replied:
What about `ossl_init_thread_start`?

mattcaswell replied:
The `stop` equivalent of `ossl_init_thread_start` is `ossl_init_thread_stop`. There is no "start" equivalent of `ossl_ctx_thread_stop`.

slontis commented on crypto/initthread.c:
```diff
@@ -0,0 +1,248 @@
+
+int init_thread(void)
+{
+
```
NIT: blankline

slontis commented on crypto/initthread.c:
```diff
@@ -0,0 +1,248 @@
+    }
+
+    if (freehands)
+        OPENSSL_free(hands);
```
Should *hands be NULL here?

slontis commented on crypto/initthread.c:
```diff
@@ -0,0 +1,248 @@
+        prev = curr;
+        curr = curr->next;
+        if (prev == *hands)
+            *hands = NULL;
```
Doesnt this *hands get returned?

mattcaswell replied:
I refactored this to make things a bit clearer.

slontis commented on crypto/initthread.c:
```diff
@@ -0,0 +1,248 @@
+    OPENSSL_CTX *ctx = arg;
+
+    /*
+     * In FIPS mode the list of THREAD_EVENT_HANDLERs unique per combination of
```
NIT is unique

slontis commented on crypto/initthread.c:
```diff
@@ -0,0 +1,248 @@
+                               &thread_event_ossl_ctx_method);
+#else
+    /*
+     * Outside of FIPS mode the list of THREAD_EVENT_HANDLERs is unque per
```
NIT: unique

slontis commented on crypto/rand/drbg_lib.c:
```diff
@@ -141,6 +147,161 @@ static int is_digest(int type)
+#endif
+
+    if (!CRYPTO_THREAD_init_local(&dgbl->private_drbg, NULL))
+        goto err1;
```
I assume all these goto's mean that CRYPTO_THREAD_cleanup_local() doesnt handle passing NULLS - looks  messy

mattcaswell replied:
Well, `dgbl->private_drbg` is not NULL if there's an error, just uninitialised - so we shouldn't attempt to use it in anyway.

richsalz replied:
Since dgbl is allocated via zalloc, and since NULL is all-bits-zero (via the sanity test), it is initialized.

mattcaswell replied:
It's the `private_drbg` field we are specifically concerned with there. And yes it will get set to 0 on create. However, when I said it was unitialised, I meant it has not been initialised via a call to `CRYPTO_THREAD_init_local`. We should not be calling `CRYPTO_THREAD_cleanup_local` on keys that have not been initialised in that way. We cannot know how `CRYPTO_THREAD_cleanup_local` would treat a key that is set to 0. It most likely would be fine but we cannot say for sure. For example in a pthreads implementation `CRYPTO_THREAD_cleanup_local` calls `pthread_key_delete` which has this description: "The pthread_key_delete() function shall delete a thread-specific data key previously returned by pthread_key_create()". It says nothing about what happens if you pass it zero data that was not created by `pthread_key_create` - so at best this behaviour is undefined.

slontis commented:
I havent gone through the whole thing yet - but it seems that a large percentage of this PR is related to threads. Maybe the rand changes should be in a seperate PR?

slontis commented:
travis errors are valid (except for the timeout ones)

mattcaswell commented:
I've attempt to respond to the comments so far and have also tried to fix the various travis issues. I've also rebased and force pushed.

I'm going to take a look to see if I can easily split this up into smaller PRs to make it easier to review.

mattcaswell commented:
I've split out various bits of this PR into other PRs. See #9038, #9039, and #9040. Once all of those are merged this PR will need to be rebased to pick up the last few commits to complete the move of the RAND code into the FIPS module. In the meantime I've set this PR to be WIP.

mattcaswell commented:
I rebased this now that #9038 has gone in, and to include the latest changes from #9039.

mattcaswell commented:
Needs #9186 before this can progress.

levitte commented on doc/man3/OPENSSL_CTX.pod:
```diff
@@ -20,7 +20,10 @@ Applications may allocate their own, but may also use C<NULL> to use
 argument.
 
-OPENSSL_CTX_new() creates a new OpenSSL library context.
+OPENSSL_CTX_new() creates a new OpenSSL library context. When a non default
```
Line break after sentence end, please

levitte commented on doc/man3/RAND_DRBG_get0_master.pod:
```diff
@@ -21,26 +27,35 @@ RAND_DRBG_get0_private
+for the given OPENSSL_CTX B<ctx>.
 
+In all the above cases the B<ctx> parameter can
+be NULL in which case the default OPENSSL_CTX is used. RAND_DRBG_get0_master(),
```
Line break after sentence end, please

levitte commented:
Please rebase then ping, I'll try to review quickly.

mattcaswell commented:
I've rebased this PR. Note the feedback comments above were for commits no longer in this PR. Note this PR now also includes the commits from #9186, so that needs to go in first before I can take this out of WIP.

mattcaswell commented:
Finally all the dependent PRs are in and I am able to take this out of WIP!

I've rebased on the latest master. Please review!

levitte commented on crypto/hmac/build.info:
```diff
@@ -1,3 +1,6 @@
         hmac.c hm_ameth.c hm_meth.c
+
+SOURCE[../../providers/fips]=\
+        hmac.c hm_meth.c
```
May I suggest this?

```
$COMMON=hmac.c hm_meth.c

LIBS=../../libcrypto
SOURCE[../../libcrypto]=$COMMON hm_ameth.c
SOURCE[../../providers/fips]=$COMMON
```

mattcaswell replied:
Ah, yes. This code was written before you introduced that capability. Fixed.

levitte commented on crypto/rand/build.info:
```diff
@@ -3,3 +3,8 @@ SOURCE[../../libcrypto]=\
+# FIPS Module
+SOURCE[../../providers/fips]=\
+        rand_lib.c rand_crng_test.c rand_win.c rand_unix.c  rand_vms.c \
+        drbg_lib.c drbg_ctr.c rand_vxworks.c drbg_hash.c drbg_hmac.c
```
SImilarly:

```
$COMMON=rand_lib.c rand_crng_test.c rand_win.c rand_unix.c  rand_vms.c \
        drbg_lib.c drbg_ctr.c rand_vxworks.c drbg_hash.c drbg_hmac.c

LIBS=../../libcrypto
SOURCE[../../libcrypto]=$COMMON randfile.c rand_err.c rand_egd.c
SOURCE[../../providers/fips]=$COMMON
```

mattcaswell replied:
Fixed.

levitte commented on crypto/hmac/hm_meth.c:
```diff
@@ -152,13 +152,16 @@ static int hmac_ctrl_str(EVP_MAC_IMPL *hctx, const char *type,
             return 0;
         return hmac_ctrl_int(hctx, EVP_MAC_CTRL_SET_MD, d);
     }
+#endif
```
This is problematic, 'cause directly fetched MDs are broken for the moment, since sizes and flags haven't been transferred.  I need to get back to the fixing PR, thanks for the reminder...

mspncp replied:
Has this request been resolved?

mattcaswell replied:
I think this isn't so much a request to do anything on this PR...rather a side note.

mattcaswell commented:
Hmm. This is strange. Travis is reporting a doc-nits failure:

````
*** ERROR: Apparent command =head1 not preceded by blank line at line 8 in file doc/internal/man3/ossl_init_thread_deregister.pod

make: *** [doc-nits] Error 1

+// MAKE DOC-NITS FAILED
````

However I don't get the above error locally (doc-nits passes)!!? However if I add a blank line at line 8 as suggested by the error message I get this locally:

````
*** WARNING: empty section in previous paragraph at line 9 in file doc/internal/man3/ossl_init_thread_deregister.pod
Makefile:524: recipe for target 'doc-nits' failed
make: *** [doc-nits] Error 1
````
In any case, neither of these issues is directly related to this PR.

mattcaswell commented:
Fixup commit pushed addressing the feedback, and also fixing a windows linking issue. I've not addressed the doc-nits issue because that is (presumably) also a problem on master so probably is better done in a different PR.

mattcaswell commented:
doc-nits issue fixed in #9192

levitte commented:
Looks good to me. I would like to know what @mspncp has to say, though.

mspncp commented:
> Looks good to me. I would like to know what @mspncp has to say, though.

I started taking a look, but it might take a while to complete it, since I'm currently on vacation.

levitte commented:
Oh, sorry, forgot about that...

mspncp commented on crypto/rand/drbg_hash.c:
```diff
@@ -303,23 +304,39 @@ static RAND_DRBG_METHOD drbg_hash_meth = {
-    md = EVP_get_digestbynid(drbg->type);
+    md = EVP_MD_meth_dup(EVP_get_digestbynid(drbg->type));
+#else
+    /* TODO(3.0): Fill this out with the complete list of allowed digests */
```
Continuing my previous comment: For the moment, I would add all algorithms available for the default provider and change the TODO comment to remove the non-fips algorithms from the switch.

mattcaswell replied:
Done

mspncp commented on crypto/rand/drbg_hmac.c:
```diff
@@ -197,23 +198,37 @@ static RAND_DRBG_METHOD drbg_hmac_meth = {
+#ifndef FIPS_MODE
     /* Any approved digest is allowed - assume we pass digest (not NID_hmac*) */
-    md = EVP_get_digestbynid(drbg->type);
+    md = EVP_MD_meth_dup(EVP_get_digestbynid(drbg->type));
```
No generic method lookup for the HMAC-DRBG either. See my comments for the HASH-DRBG.

mattcaswell replied:
Done

mspncp commented on crypto/rand/drbg_hmac.c:
```diff
@@ -197,23 +198,37 @@ static RAND_DRBG_METHOD drbg_hmac_meth = {
-    md = EVP_get_digestbynid(drbg->type);
+    md = EVP_MD_meth_dup(EVP_get_digestbynid(drbg->type));
+#else
+    /* TODO(3.0): Fill this out with the complete list of allowed digests */
```
See my comments for the HASH-DRBG.

mattcaswell replied:
Done

mattcaswell commented on crypto/hmac/hm_meth.c:
```diff
@@ -152,13 +152,16 @@ static int hmac_ctrl_str(EVP_MAC_IMPL *hctx, const char *type,
     if (!value)
         return 0;
+#ifndef FIPS_MODE
+    /* Not supported in FIPS mode due to the implict fetch of the md */
```
"Explicit" fetch occurs when we look up an algorithm using one of the "fetch" functions such as EVP_MD_fetch(). This finds the algorithm and associates it with an implementation from a provider. Conversely "Implicit" fetch is when we don't do that, and use a function such as "EVP_sha256()" which has no provider implementation associated with it. With an algorithm obtained in that way the "fetch" happens automatically when you try to use it - hence "implicit fetch". An important difference between these two types of EVP_MD objects is that explicitly fetched versions are dynamically allocated and must be "freed", whereas implicitly fetched ones do not. That fact makes this code more complicated if we want to convert it to use explicit fetch. Since we are unlikely to need this ctrl in FIPS_MODE I didn't bother.

I replaced this comment with a longer one to try and explain this better.

mattcaswell commented on crypto/rand/drbg_hash.c:
```diff
@@ -303,23 +304,39 @@ static RAND_DRBG_METHOD drbg_hash_meth = {
+#ifndef FIPS_MODE
     /* Any approved digest is allowed */
-    md = EVP_get_digestbynid(drbg->type);
+    md = EVP_MD_meth_dup(EVP_get_digestbynid(drbg->type));
```
> I don't think the digest algorithm of the HASH-DRBG should be selectable via generic EVP lookup accross multiple providers
...
> The aes-ctr algorithm is always fetched unconditionally from the provider context.

I think there is some confusion here. Ignoring the FIPS provider for a moment (which is a special case), the DRBG code exists in libcrypto and is not associated with a particular provider (yet). We probably will want to make it possible to "fetch" a DRBG implementation from a provider and use that...but we're not there yet. For now its just in libcrypto. The look up across "multiple providers" will happen anyway when used, even with aes-ctr. It just looks for an implementation of (for example) "AES-128-ECB" and uses it. The default provider (which might not even be available for all we know) may or may not be used.

Now, as I said, the FIPS provider is a special case. It does not link against libcrypto and so can't use the built-in DRBG implementation. Therefore, for that provider only, we are additionally making the DRBG code available inside the FIPS provider. For that provider only, when we do a "fetch" it will only ever find algorithms provided by itself.

Anyway, aside from that confusion, your suggestion is fine and I have refactored things accordingly.

paulidale replied:
I'd lean towards making the digest selection completely open.  It provides a degree of flexibility in case something untoward happens to one of the digests.  However, I'm happy enough with the current switch statement to not be too bothered (there are plenty of choices available and it avoids an explicit XOF check).

One could argue that the CTR-DRBG should have this flexibility too (ARIA and Camellia should both drop in instead of AES).  Still, for some point in the future rather than now.

mspncp replied:
I don't mind, as long as the implementation is consistent for the three DRBG implementations. Anyway, if this change is desired it would be a matter for a separate pull request.

mattcaswell commented:
New commits addressing the feedback pushed.

mspncp commented on crypto/rand/drbg_hash.c:
```diff
@@ -303,23 +305,55 @@ static RAND_DRBG_METHOD drbg_hash_meth = {
+#ifndef FIPS_MODE
+    case NID_blake2b512:
+    case NID_blake2s256:
+    case NID_sm3:
```
This is more a question to @paulidale and @slontis than a review comment:  [NIST SP 800-90Ar1](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-90Ar1.pdf) does not mention SHA3, BLAKE and SM3. Is it 'legal' nevertheless to use them?

paulidale replied:
Section 10.1 opens by saying that the hash DRBG mechanisms _have been designed to use any approved hash function_.  Based on this SHA3 **would** be legal.

BLAKE and SM3 cannot be FIPS validated.

mspncp commented on crypto/rand/drbg_hmac.c:
```diff
@@ -197,23 +199,53 @@ static RAND_DRBG_METHOD drbg_hmac_meth = {
+#ifndef FIPS_MODE
+    case NID_blake2b512:
+    case NID_blake2s256:
+    case NID_sm3:
```
Same question to @paulidale and @slontis.

paulidale replied:
I think it has the same answer.  10.1.2 again mentions an _approved_ hash function.

mspncp commented on crypto/rand/rand_crng_test.c:
```diff
@@ -64,27 +65,29 @@ static const OPENSSL_CTX_METHOD rand_crng_ossl_ctx_method = {
 };
 
 int rand_crngt_get_entropy_cb(OPENSSL_CTX *ctx,
+                              RAND_POOL *pool,
```
What is the rationale behind the changes in `rand_crng_test.c`?

mspncp replied:
Is it a fixup for 4e297b7441a0 which you made along the way?

mattcaswell replied:
Effectively yes. I encountered a crash related to this code. The change is related to the `crngt_get_entropy` callback. That callback gets called from `rand_crng_ossl_ctx_new`. That function is called when we are initialising an OPENSSL_CTX with CRNG data for the first time. Unfortunately, in the original implementation, the default callback was itself trying to get CRNG data from the OPENSSL_CTX. But since we're still initialising it it went round in an an infinite recursive loop trying to initialise the OPENSSL_CTX until eventually it blew the stack. The solution to the problem was to change the callback to take the RAND_POOL to be used as a parameter instead of getting it from the OPENSSL_CTX.

mspncp commented on crypto/rand/rand_win.c:
```diff
@@ -162,7 +162,7 @@ int rand_pool_add_additional_data(RAND_POOL *pool)
 }
 
-# if !OPENSSL_API_1_1_0
+# if !OPENSSL_API_1_1_0 || defined(FIPS_MODE)
```
Are you sure to enable it in FIPS mode? Or did you mean
```
# if !OPENSSL_API_1_1_0 && !defined(FIPS_MODE)
```

mattcaswell replied:
Yes, I did mean that!

mspncp replied:
Hmmm... but why? The code is marked deprecated-in-1.1.1 and not called anywhere.

```
msp@msppc:~/src/openssl$ git grep -E -nH 'RAND_(event|screen)' pr-9035
pr-9035:crypto/rand/rand_win.c:166:int RAND_event(UINT iMsg, WPARAM wParam, LPARAM lParam)
pr-9035:crypto/rand/rand_win.c:172:void RAND_screen(void)
pr-9035:doc/man3/RAND_add.pod:5:RAND_add, RAND_poll, RAND_seed, RAND_status, RAND_event, RAND_screen,
pr-9035:doc/man3/RAND_add.pod:25: int RAND_event(UINT iMsg, WPARAM wParam, LPARAM lParam);
pr-9035:doc/man3/RAND_add.pod:26: void RAND_screen(void);
pr-9035:doc/man3/RAND_add.pod:74:RAND_event() and RAND_screen() are equivalent to RAND_poll() and exist
pr-9035:doc/man3/RAND_add.pod:84:RAND_event() returns RAND_status().
pr-9035:doc/man3/RAND_add.pod:98:RAND_event() and RAND_screen() were deprecated in OpenSSL 1.1.0 and should
pr-9035:include/openssl/rand.h:68:DEPRECATEDIN_1_1_0(void RAND_screen(void))
pr-9035:include/openssl/rand.h:69:DEPRECATEDIN_1_1_0(int RAND_event(UINT, WPARAM, LPARAM))
pr-9035:util/libcrypto.num:1331:RAND_event                              1318    3_0_0   EXIST:_WIN32:FUNCTION:DEPRECATEDIN_1_1_0
pr-9035:util/libcrypto.num:1816:RAND_screen                             1804    3_0_0   EXIST:_WIN32:FUNCTION:DEPRECATEDIN_1_1_0
```

mspncp replied:
Is it because it is still referenced in libcrypto.num?

```
pr-9035:util/libcrypto.num:1331:RAND_event                              1318    3_0_0   EXIST:_WIN32:FUNCTION:DEPRECATEDIN_1_1_0
pr-9035:util/libcrypto.num:1816:RAND_screen                             1804    3_0_0   EXIST:_WIN32:FUNCTION:DEPRECATEDIN_1_1_0
```

mattcaswell replied:
Sorry, I wasn't clear. What *you* wrote is what I meant! I've fixed it now.

mspncp replied:
But the FIPS provider does not reuse libcrypto.num, right?

mspncp replied:
> Sorry, I wasn't clear. What _you_ wrote is what I meant! I've fixed it now.

Ok, now the world is in order again :-)

mattcaswell replied:
> But the FIPS provider does not reuse libcrypto.num, right?

Right. It has its own num file which exports exactly one function:

https://github.com/openssl/openssl/blob/master/util/providers.num

mattcaswell commented:
> BTW: Is it actually possible by now to run and check the DRBG code of the FIPS provider in the debugger (on linux)? If yes, what do I need to set it up?

With this PR, yes. Set a breakpoint on line 120 of `providers/fips/fipsprov.c` and then run `evp_extra_test`:

````
$ OPENSSL_MODULES=providers util/shlib_wrap.sh gdb --args test/evp_extra_test -test 13
GNU gdb (Ubuntu 8.1-0ubuntu3) 8.1.0.20180409-git
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from test/evp_extra_test...done.
(gdb) break fipsprov.c:120
No source file named fipsprov.c.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (fipsprov.c:120) pending.
(gdb) run
Starting program: /home/matt/dev/openssl-write/test/evp_extra_test -test 13
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
    # Subtest: test_EVP_MD_fetch
    1..5
    ok 1 - iteration 1
    ok 2 - iteration 2
    ok 3 - iteration 3

Breakpoint 1, dummy_evp_call (provctx=0x5555557c0530)
    at providers/fips/fipsprov.c:120
120	    if (RAND_DRBG_bytes(drbg, randbuf, sizeof(randbuf)) <= 0)
````

mspncp commented:
Thanks for the instructions. I'll give it a try this evening...

mattcaswell commented on doc/internal/man3/ossl_prov_util_nid_to_name.pod:
```diff
@@ -0,0 +1,35 @@
+
+ #include "internal/providercommon.h"
+
+ const char *ossl_prov_util_nid_to_name(int nid);
```
@levitte has been championing the documentation of the "internal" API, i.e. stuff that ends up in our internal header files. There's quite a few man pages there already. This particular function is purely internal. There is no intention for it to be used by third party providers.

mattcaswell commented on crypto/hmac/hm_meth.c:
```diff
@@ -152,13 +152,23 @@ static int hmac_ctrl_str(EVP_MAC_IMPL *hctx, const char *type,
+    /*
+     * We don't have EVP_get_digestbyname() in FIPS_MODE. That function returns
+     * an EVP_MD without an associated provider implementation (i.e. it is
+     * using "implict fetch"). We could replace it with an "explicit" fetch
```
Fixed.

mattcaswell commented:
Fix up commit pushed addressing the feedback.

mattcaswell commented:
>  I'll do a live examination in the debugger this evening.

Did you manage to do that?

mspncp commented:
Yes I did, and I have some thoughts to share, which I did not find the time to write down yet. Also, I might have some questions for you. But it's nothing that would prevent this pull request from being merged, I haven't found any issues (yet). I'd still be more comfortable if  @levitte would approve the provider related (non-drbg) parts of this pull request, but if you think my approval is sufficient, feel free to go ahead and merge it.

mattcaswell commented:
> I'd still be more comfortable if @levitte would approve the provider related (non-drbg) parts of this pull request, but if you think my approval is sufficient, feel free to go ahead and merge it.

I think @levitte is away at the moment with only occasional network access. I'll leave this open a little bit longer to see if he comments, but otherwise I will merge this later today.

mattcaswell commented:
Pushed! Thanks!
